src/ejabberd_c2s_info_handler.erl

```diff
@@ -0,0 +1,56 @@
+-module(ejabberd_c2s_info_handler).
+
+-include("ejabberd_c2s.hrl").
+-include("mongoose.hrl").
+
+% API for other modules
+-export([add/5, remove/2]).
+
+% API for ejabberd_c2s
+-export([get_for/2,
+         add_to_state/5,
+         remove_from_state/2,
+         safe_call/3,
+         safe_call/6]).
+
+add(C2S, Tag, M, F, HandlerState) ->
+    p1_fsm_old:send_all_state_event(C2S, {add_info_handler, Tag, M, F, HandlerState}).
+
+remove(C2S, Tag) ->
+    p1_fsm_old:send_all_state_event(C2S, {remove_info_handler, Tag}).
+
+get_for(Tag, #state{ info_handlers = InfoHandlers }) ->
+    maps:get(Tag, InfoHandlers, undefined).
```

Comment:
I'd not return `undefined` from any function whatsoever, as it makes troubleshooting a pain in the neck. The `undefined` weaves through following code and makes it hard to debug, when it actually a crash. [`undefined` is the `NULL` of Erlang](https://www.lucidchart.com/techblog/2015/08/31/the-worst-mistake-of-computer-science/).  Instead, I think the `maps` module has the sanest policy for fetching values:

- there's either `maps:get(Key, Map)` which throws an exception if `Key` is not found
- or `maps:get(Key, Map, Default)` which returns `Default` in case `Key` is not in `Map`

this allows us to either fail-fast and catch the exception or get a useful default if the failure is not critical and we know what to do if without the value. Returning `undefined` is the worst possible case as we neither fail-fast, nor have a usable value to deal with in subsequent code.

Reply:
I fully agree with you @erszcz.
In this particular case I thought about using `maps:find` function which returs `{ok, Value}` or `error`. This is not ideal as well since the `error` can be considered as harmful as `undefined`.

Reply:
I disagree. It doesn't matter whether it's `error`, `undefined` or some other atom returned. In case of this function the contract is that it either returns 3-element tuple or `undefined`. Perhaps it may be renamed to `find_for` to be consistent with e.g. maps or lists API.
I understand the concern behind `undefined` being Erlang's `NULL` but the alternative is to crash and use `try` in `c2s`. Replacing all `undefined` with `try` blocks sounds like making Erlang another Java.

Reply:
@fenek In places where you don't want to crash/`catch`, use the variant which accepts the `Default`. That's why the proposed API is the most flexible - it allows you to follow any path you wish.

Reply:
The whole purpose of such an API is to write *less* code in any case.

Reply:
In yet other words, an access function should have _fail fast_ crashing variant and a non-crashing variant.

For the second one it could return `undefined`, but then the handling code gets clunky, as we have to `case ... of` around it. The best design is to make the function itself take a default  value and return it, as that way we move the `case` inside the getter instead of the call site.